### PR TITLE
abstract number parsing

### PR DESCRIPTION
**Pull request by anqit**
This is a step towards allowing all of the incredible number formatting work that has been done in this package to be reused for more general or custom purposes. For example, [this feature request](https://github.com/a14n/dart-decimal/issues/52) could then be fairly trivial to implement. 

Sample use case: https://github.com/anqit/dart-decimal-number-parser, happy to hear better ideas to accomplish something similar.

---

- [X] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/wiki/External-Package-Maintenance#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

### CODE REVIEW

**Comment by anqit** [comment]
@mosuem I see you've approved some other PRs, is there any way to get some thoughts on this?

File: pkgs/intl/lib/src/intl/number_format.dart [hunk 1]
```diff
@@ -495,7 +496,13 @@ class NumberFormat {
 
   /// Parse the number represented by the string. If it's not
   /// parseable, throws a [FormatException].
-  num parse(String text) => NumberParser(this, text).value!;
+  // TODO: use constructor tear-off syntax when dart sdk version >= 2.15
+  num parse(String text) => parseWith(text, (f, t) => NumberParser(f, t));
```
Comment (mosuem) [inline pkgs/intl/lib/src/intl/number_format.dart:500]: Nit: 
```suggestion
  num parse(String text) => parseWith((f, t) => NumberParser(f, t), text);
```

File: pkgs/intl/lib/src/intl/number_format.dart [hunk 2]
```diff
@@ -495,7 +496,13 @@ class NumberFormat {
 
   /// Parse the number represented by the string. If it's not
   /// parseable, throws a [FormatException].
-  num parse(String text) => NumberParser(this, text).value!;
+  // TODO: use constructor tear-off syntax when dart sdk version >= 2.15
```
Comment (mosuem) [inline pkgs/intl/lib/src/intl/number_format.dart:499]: You can also just increase the SDK constraint to allow tear-offs.

File: pkgs/intl/lib/src/intl/number_parser.dart [hunk 3]
```diff
@@ -2,240 +2,31 @@
...
 
-import 'package:intl/number_symbols.dart';
+import 'package:intl/src/intl/number_parser_base.dart';
 
-import 'constants.dart' as constants;
 import 'number_format.dart';
-import 'number_format_parser.dart';
-import 'string_stack.dart';
 
 ///  A one-time object for parsing a particular numeric string. One-time here
```
Comment (mosuem) [inline pkgs/intl/lib/src/intl/number_parser.dart:9]:
```suggestion
/// A one-time object for parsing a particular numeric string. One-time here
```

**Review by mosuem — CHANGES_REQUESTED** [review]

**Comment by mosuem** [comment]
Some small changes, otherwise LGTM! Sorry about the delay.

**Comment by anqit** [comment]
> Some small changes, otherwise LGTM! Sorry about the delay.

Np, @mosuem, thanks for taking a look. I believe I've addressed your comments.

File: pkgs/intl/lib/src/intl/number_format.dart (hunk 2 above)
Comment (mosuem) [reply pkgs/intl/lib/src/intl/number_format.dart]: The sdk constraint is `sdk: ^3.0.0` now anyways, so feel free to remove the todo.

**Review by mosuem — APPROVED** [review]

**Comment by anqit** [comment]
@mosuem can this be merged?

**Comment by anqit** [comment]
@mosuem curious what kind of release cadence, if any, this library follows? any idea when a 0.19 release could be expected? thanks!

**Pull request merged**